### PR TITLE
MAGENTO-2306 refactoring wishlist ID getter + storage item removal

### PR DESCRIPTION
**Pull request by augsteyer**
## Description

Performance update for wishlist ID getting. Added removal of wishlist products for future performance gains. No change log as the initial version is not released.

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I updated the CHANGELOG.md

### CODE REVIEW

File: extension/lib/magento/customer/deleteItems.js [hunk 1]
```diff
@@ -12,11 +12,19 @@ module.exports = async (context, input) => {
   const wishlistItemIdsString = input.productIds.map(productId => wishlistItemIdMapping[productId]).toString()
 
   if (!wishlistItemIdsString) {
-    return
+    context.log.debug('No products of this mapping exist in the storage.')
+    return deleteItemsFromStorage()
```
Comment (Skistims) [inline extension/lib/magento/customer/deleteItems.js]: if the wishlistItemIdsString is empty, it means the productIds were not found in the mapping from the storage. This call seems unnecessary

File: extension/lib/magento/customer/deleteItems.js [hunk 2]
```diff
@@ -12,11 +12,19 @@ module.exports = async (context, input) => {
...
   }
 
   const request = new MagentoRequest(context, input.token)
   const itemFilter = `?wishlistItemIds=${wishlistItemIdsString}`
   const wishlistItemsEndpointUrl = `${context.config.magentoUrl}/wishlists/${input.wishlistId}/items${itemFilter}`
-  request.send(wishlistItemsEndpointUrl, 'deleteFavorites', 'DELETE')
+  await request.send(wishlistItemsEndpointUrl, 'deleteFavorites', 'DELETE')
+  await deleteItemsFromStorage()
```
Comment (Skistims) [inline extension/lib/magento/customer/deleteItems.js]: definitely no need to await this.

File: extension/lib/magento/customer/deleteItems.js [hunk 3]
```diff
@@ -12,11 +12,19 @@ module.exports = async (context, input) => {
...
-    return
+    context.log.debug('No products of this mapping exist in the storage.')
+    return deleteItemsFromStorage()
   }
 
   const request = new MagentoRequest(context, input.token)
   const itemFilter = `?wishlistItemIds=${wishlistItemIdsString}`
   const wishlistItemsEndpointUrl = `${context.config.magentoUrl}/wishlists/${input.wishlistId}/items${itemFilter}`
-  request.send(wishlistItemsEndpointUrl, 'deleteFavorites', 'DELETE')
+  await request.send(wishlistItemsEndpointUrl, 'deleteFavorites', 'DELETE')
```
Comment (Skistims) [inline extension/lib/magento/customer/deleteItems.js:25]: probably no need to await this.
As we do not need reponse (it is empty anyway) and the errorhandling is happening within the send function.
So waiting for the API here causes unnecessary delay

File: extension/lib/magento/customer/deleteItems.js [hunk 4]
```diff
@@ -12,11 +12,19 @@ module.exports = async (context, input) => {
...
+  await deleteItemsFromStorage()
+
+  async function deleteItemsFromStorage () {
+    input.productIds.map(id => {
+      context.storage.user.map.delItem('wishlistItemIdMapping', id)
```
Comment (Skistims) [inline extension/lib/magento/customer/deleteItems.js]: for the record:
If you want to keep the item mapping clean all the time (no dead entries) that is okay, even though I don't see any upside.
Just wanted to let you know, that the mapping is reinitiated every time the getFavorites pipeline is called. This is also implied, when the putFavorites pipeline is called, which happens pretty often.
So maybe this changes your mind about cleaning this up at all.

**Review by Skistims — CHANGES_REQUESTED** [review]

File: extension/lib/magento/customer/deleteItems.js (hunk 3 above)
Comment (augsteyer) [reply extension/lib/magento/customer/deleteItems.js:25]: will have to confirm on how contained the `steps` are, aka can 2 steps be getting called at the same time

File: extension/lib/magento/customer/deleteItems.js (hunk 4 above)
Comment (augsteyer) [reply extension/lib/magento/customer/deleteItems.js]: yup, did not notice that due to how inconsistently the storage clearing worked, see the **possible** fix. I say possible as it is hard to be a 100% with async bugs due to inconsistent results with different server loads/runs.

**Review by Skistims — APPROVED** [review]

File: extension/lib/magento/customer/deleteItems.js (hunk 3 above)
Comment (augsteyer) [reply extension/lib/magento/customer/deleteItems.js:25]: Waldemar confirmed that there can be issues between steps and async calls. We do not want to create a race condition between the delete endpoint call and the get wishlist item call. So if for some reason magento hiccups and delete takes longer than the getItems, then the getItems call will think that there are more products than there actually are.

**Pull request merged**